### PR TITLE
Attacher improvements

### LINKED ISSUE #1688: RemoteAttacher: control log level
Hello!

## Is your feature request related to a problem?

I'm working at Clever Cloud and we provide your agent to our customers.  When deploying Java applications, we pass the full environment variables of the application using the `-D` flag of java. But then, the `RemoteAttacher` will log the whole Java command line used to start the JVM, including the environment variable values which is then collected into our log system.

https://github.com/elastic/apm-agent-java/blob/8545eb2be19cea7139801e78f7704dd7c4693f7f/apm-agent-attach/src/main/java/co/elastic/apm/attach/RemoteAttacher.java#L130

The `ELASTIC_APM_LOG_LEVEL` environment variable doesn't control the log level of the attacher, only of the agent once injected. The `log()` method doesn't really filter logs based on their level: https://github.com/elastic/apm-agent-java/blob/8545eb2be19cea7139801e78f7704dd7c4693f7f/apm-agent-attach/src/main/java/co/elastic/apm/attach/RemoteAttacher.java#L98...L100

## Describe the solution you'd like

Would it be possible to add a log level control, even very basic, for this? Maybe one that respects `ELASTIC_APM_LOG_LEVEL`?

## Describe alternatives you've considered

One alternative could be that we (Clever Cloud) could hide those logs ourselves but it may be better to be done upstream. I can take some time to make a PR if you're willing to have such a feature.

### PR DESCRIPTION
**Pull request by felixbarny**
## What does this PR do?

From changelog.asciidoc:

* Overhaul of the <<setup-attach-cli,attacher cli>> application that allows to attach the agent to running JVMS
  * The artifact of the standalone cli applicaiton is now called `apm-agent-attach-cli`. The attacher API is still called `apm-agent-attach`.
  * There is also a slim version of the cli application that does not bundle the Java agent.
   It requires the `--agent-jar` option to be set.
  * Improved logging
   The application uses {ecs-logging-java-ref}/intro.html[Java ECS logging] to emit JSON logs.
   The log level can be configured with the `--log-level` option.
   By default, the program is logging to the console but using the `--log-file` option, it can also log to a file.
  * Attach to JVMs running under a different user (unix only)
   The JVM requires the attacher to be running under the same user as the target VM (the attachee).
   The `apm-agent-attach-standalone.jar` can now be run with a user that has permissions to switch to the user that runs the target VM.
   On Windows, the attacher can still only attach to JVMs that are running with under the same user.
  * New include/exclude discovery rules
    * `--include-all`: Attach to all discovered JVMs. If no matchers are provided, it will not attach to any JVMs.
    * `--include-user`/`--exclude-user`: Attach to all JVMs of a given operating system user.
    * `--include-main`/`--exclude-main`: Attach to all JVMs that whose main class/jar name, or system properties match the provided regex.
    * `--include-vmargs`/`--exclude-vmargs`: Attach to all JVMs that whose main class/jar name, or system properties match the provided regex.
  * Removal of options
    * The deprecated `--arg` option has been removed.
    * The `-i`/`--include`, `-e`/`exclude` options have been removed in favor of the `--<include|exclude>-<main|vmargs>` options.
    * The `-p`/`--pid` options have been removed in favor of the `--include-pid` option.
  * Changed behavior of  the `-l`/`--list` option
   The option now only lists JVMs that match the include/exclude discovery rules.
   Thus, it can be used to do a dry-run of the matchers without actually performing an attachment.
   It even works in combination with `--continuous` now.
   By default, the VM arguments are not printed, but only when the `-a`/`--list-vmargs` option is set.
  * Remove dependency on `jps`
   Even when matching on the main class name or on system properties,
  * Checks the Java version before attaching to avoid attachment on unsupported JVMs.

## Checklist

- [ ] This is an enhancement of existing features, or a new feature in existing plugins
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [ ] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)

TODOs
- [x] The `apm-agent-attach.jar` (attacher API) now has a compile dependency to log4j2 and log4j2-ecs-layout. Maybe we'll have to create different modules for the standalone cli attacher and the attacher API.
- [x] allow to log to a file
- [x] Allow specifying the path to the agent jar. Consider publishing an attacher that doesn't bundle the agent jar.
- [x] Rethink include/exclude rule logic. Currently, all conditions are applied with an OR. Example: `--include-cmd foo --exclude-cmd bar` would match `foo-bar.jar`. Should a given JVM pass the filter if it matches any include rule AND not matches any exclude rule?
    For Elastic Agent, they have recently added a similar rule logic. The following only allows APM to run:

```yaml
constraints:
- rule: allow
  input: apm
- rule: deny
  input: "*"
```

## Related issues
Closes #1688

### CODE REVIEW

**Comment by SylvainJuge** [comment]
> --pid is deprecated in favor of --include-pid
Seems to contradict
> The -p/--pid options have been removed in favor of the --include-pid option.

Given the number of changes here, breaking compatibility with the `--pid` option could be a relevant option to ensure that other changes are not silently ignored on upgrade.

Also, most of the support cases we had in the past with `--include/--exclude` were due to the fact that it was relying on `jps` output parsing which isn't the most straightforward.

What exactly would be the use-case of the attacher without bundled agent in practice ? Having them bundled together tends to simplify deployment and compatibility checks.

**Comment by felixbarny** [comment]
> --pid is deprecated in favor of --include-pid

That's outdated. I removed -p/--pid.

> What exactly would be the use-case of the attacher without bundled agent in practice ? Having them bundled together tends to simplify deployment and compatibility checks.

I think that this is the mode we'll be using when integrating APM Server. While we want to be able to update the Java agent independently from APM Server, there is a strong dependency between the UI, the server and the attacher-cli. These three components have to have a known set of features. If we were to implement a new type of include/exclude matcher in the attacher-cli in the future, we couldn't easily expose it in the UI if different versions of the attacher-cli may be used.

So, we'd bundle the attacher-cli-slim into APM Server and download the agent jar on demand, based on what's configured in the Fleet UI.

**Comment by felixbarny** [comment]
Note that there's still the non-slim version of the attacher-cli that does ship with a bundled agent jar.

**Review by SylvainJuge — APPROVED** [review]
LGTM, only few comments/questions.
Will definitely make attachment easier, thus a breaking change is welcome for once !

File: apm-agent-attach-cli/src/main/java/co/elastic/apm/attach/JvmDiscoverer.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        private final UserRegistry userRegistry;
+
+        public ForHotSpotVm(List<String> tempDirs, UserRegistry userRegistry) {
+            this.tempDirs = tempDirs;
+            this.userRegistry = userRegistry;
+        }
+
+        public static ForHotSpotVm withDiscoveredTempDirs(UserRegistry userRegistry) {
+            List<String> tempDirs = new ArrayList<>();
+            if (Platform.isMac()) {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-attach-cli/src/main/java/co/elastic/apm/attach/JvmDiscoverer.java:93]: `UserRegistry.getAllUsersMacOs().getAllTempDirs()` ends up calling `UserRegistry#findTempDir`. Turns out this is the only method that ends up calling `UserRegistry#findTempDir`. That means we can remove the fallback in `UserRegistry#findTempDir` and only rely on `java.io.tmpdir`.

File: integration-tests/application-server-integration-tests/src/test/java/co/elastic/apm/servlet/AbstractServletContainerIntegrationTest.java [hunk 2]
```diff
@@ -157,7 +157,7 @@ protected AbstractServletContainerIntegrationTest(GenericContainer<?> servletCon
             .withLogConsumer(new StandardOutLogConsumer().withPrefix(containerName))
             .withExposedPorts(webPort)
             .withFileSystemBind(pathToJavaagent, "/elastic-apm-agent.jar")
-            .withFileSystemBind(pathToAttach, "/apm-agent-attach-standalone.jar")
+            .withFileSystemBind(pathToAttach, "/apm-agent-attach-cli.jar")
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply integration-tests/application-server-integration-tests/src/test/java/co/elastic/apm/servlet/AbstractServletContainerIntegrationTest.java:160]: All artifacts get deployed, unless they opt-out by setting `<maven-deploy-plugin.skip>true</maven-deploy-plugin.skip>`

**Pull request merged**